include/exadg/operators/operator_base.cpp

```diff
@@ -2153,6 +2157,175 @@ OperatorBase<dim, Number, n_components>::evaluate_face_integrals() const
+    std::vector<std::vector<dealii::types::global_dof_index>> dof_indices_all_cells(
+      n_cells, std::vector<dealii::types::global_dof_index>(dofs_per_cell));
+    // and compute weights by counting the contributions to a DoF
+    initialize_dof_vector(weights);
+    for(unsigned int cell_batch_index = 0; cell_batch_index < matrix_free->n_cell_batches();
+        ++cell_batch_index)
+    {
+      unsigned int const n_filled_lanes =
+        matrix_free->n_active_entries_per_cell_batch(cell_batch_index);
+
+      for(unsigned int v = 0; v < n_filled_lanes; ++v)
+      {
+        auto const & cell = matrix_free->get_cell_iterator(cell_batch_index, v);
+
+        auto & dof_indices = dof_indices_all_cells[cell_batch_index * vectorization_length + v];
+        if(is_mg)
+          cell->get_mg_dof_indices(dof_indices);
+        else
+          cell->get_dof_indices(dof_indices);
+
+        for(auto const & i : dof_indices)
+          weights[i] += 1.;
+      }
+    }
+    weights.compress(dealii::VectorOperation::add);
+
+    // prepare the weights vector for symmetric weighting
+    for(unsigned int i = 0; i < weights.size(); ++i)
+    {
+      if(weights.in_local_range(i))
+        weights[i] = 1. / std::sqrt(weights[i]);
+    }
+    weights.update_ghost_values();
+
+    // cut out overlapped block matrices
+    std::vector<dealii::FullMatrix<Number>> overlapped_cell_matrices(
+      n_cells, dealii::FullMatrix<Number>(dofs_per_cell));
+
+    dealii::SparseMatrixTools::restrict_to_full_matrices<SparseMatrix,
+                                                         dealii::DynamicSparsityPattern,
+                                                         Number>(tmp_matrix,
+                                                                 dsp,
+                                                                 dof_indices_all_cells,
+                                                                 overlapped_cell_matrices);
+
+    // factorize and store cell matrices
+    matrices.resize(n_cells, LAPACKMatrix(dofs_per_cell));
+    for(unsigned int cell_batch_index = 0; cell_batch_index < matrix_free->n_cell_batches();
+        ++cell_batch_index)
+    {
+      unsigned int const n_filled_lanes =
+        matrix_free->n_active_entries_per_cell_batch(cell_batch_index);
+
+      for(unsigned int v = 0; v < n_filled_lanes; ++v)
+      {
+        // get overlapped cell matrix
+        auto const & overlapped_cell_matrix =
+          overlapped_cell_matrices[cell_batch_index * vectorization_length + v];
+
+        // store the cell matrix and renumber lexicographic
+        auto & lapack_matrix = matrices[cell_batch_index * vectorization_length + v];
+
+        auto const & lex = matrix_free->get_shape_info().lexicographic_numbering;
+        for(unsigned int i = 0; i < dofs_per_cell; i++)
+          for(unsigned int j = 0; j < dofs_per_cell; j++)
+            lapack_matrix.set(i, j, overlapped_cell_matrix[lex[i]][lex[j]]);
```

Comment:
`FE_Q` is numbered hierarchical but `FEEvaluation` expects lexicographic numbering. Should we add the same reference like here https://github.com/exadg/exadg/blob/ee073c7979b3c695aca00ade2ab9ecfebc74a410/include/exadg/operators/operator_base.cpp#L1894-L1902 ?

`FEEvaluation::get_dof_indices()` would indeed sometimes make sense, but we usually don't access the dof indices in performance-critical places.

We could also write this as a cell loop but I tend to write out the for loop over cell batches when we don't work on dof vectors.

Reply:
I am somewhat confused. according to your explanations and https://dealii.org/current/doxygen/deal.II/structinternal_1_1MatrixFreeFunctions_1_1ShapeInfo.html#abad0471d12c4f07f3d83271377e24231, I would expect the code to read
```c++
lapack_matrix.set(lex[i], lex[j], overlapped_cell_matrix[i][j]);
```
because `lapack_matrix` uses lexicographic numerbing  and `overlapped_cell_matrix` hierarchical numbering, and because shape_info::lexicographic_numbering is a  "Renumbering from deal.II's numbering of cell degrees of freedom to lexicographic numbering used inside FEEvaluation"?

 Generally, I think one should be able to get dof indices in the meaning of / consistent with `integrator.read/begin_dof_values()`.

Reply:
I discussed this briefly with @peterrum. We think both the name `shape_info.lexicograhpic_numbering` and the deal.II docu is confusing. It should be `lexicographic_to_hierarchical_numbering` in my opinion.

Reply:
The documentation is really confusing, I agree. The renumbering in correct as it currently is.

Reply:
I renamed the variable but the docu has to be fixed in deal.II. 13e5d649da7b97d9c72ab7001e949634bc32c0eb